internal/balancergroup/balancergroup.go

```diff
@@ -298,17 +298,29 @@ func (bg *BalancerGroup) Start() {
 	bg.outgoingMu.Unlock()
 }
 
-// Add adds a balancer built by builder to the group, with given id.
-func (bg *BalancerGroup) Add(id string, builder balancer.Builder) {
+// AddWithClientConn adds a balancer with the given id to the group. The
+// balancer is built with a balancer builder registered with balancerName. The
+// given ClientConn is passed to the newly built balancer instead of the
+// onepassed to balancergroup.New().
+//
+// TODO: Get rid of the existing Add() API and replace it with this.
+func (bg *BalancerGroup) AddWithClientConn(id, balancerName string, cc balancer.ClientConn) error {
+	bg.logger.Infof("Adding child policy of type %q for locality %q", balancerName, id)
+	builder := balancer.Get(balancerName)
+	if builder == nil {
+		return fmt.Errorf("unregistered balancer name %q", balancerName)
+	}
+
 	// Store data in static map, and then check to see if bg is started.
 	bg.outgoingMu.Lock()
+	defer bg.outgoingMu.Unlock()
 	var sbc *subBalancerWrapper
 	// If outgoingStarted is true, search in the cache. Otherwise, cache is
 	// guaranteed to be empty, searching is unnecessary.
 	if bg.outgoingStarted {
 		if old, ok := bg.balancerCache.Remove(id); ok {
 			sbc, _ = old.(*subBalancerWrapper)
-			if sbc != nil && sbc.builder != builder {
+			if sbc != nil && sbc.builder.Name() != builder.Name() {
```

Comment:
The pointer comparison should be fine here, right?  Especially if both were pulled from the registry?

String comparisons are slower, so it's probably better to still use the pointer, but this is pretty minor.

Reply:
I thought about this, and since we are planning to change the `balancergroup` API to take the name instead of a `balancer.Builder`, I thought name comparison was still OK here instead of the pointer. Do you have any scenarios in mind where things can go south if we do name comparisons until we switch the `balancergroup` API?

Reply:
It _shouldn't_ matter, as the balancer _should_ always come from the registry, and if it doesn't, it should also have a unique name for the builder instance.  I'm not sure if we're following that in all places where we use it, but my guess is the last item is the one least likely to be followed, which would also break this version of the code.  The pointer comparison is more optimized and results in simpler code (possibly? it's shorter, anyway), so I do think it would be better to keep it how it was.

Reply:
Reverted.